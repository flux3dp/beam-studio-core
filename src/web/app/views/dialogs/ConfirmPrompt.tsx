import classNames from 'classnames';
import React from 'react';

import ButtonGroup from 'app/widgets/ButtonGroup';
import i18n from 'helpers/i18n';
import keyCodeConstants from 'app/constants/keycode-constants';
import Modal from 'app/widgets/Modal';

const LANG = i18n.lang.alert;

interface Props {
  caption: string;
  message: string;
  confirmValue: string;
  onConfirmed: () => void;
  onClose: () => void;
  onCancel: () => void;
}

function ConfirmPrompt({
  caption, message, confirmValue,
  onCancel, onClose, onConfirmed,
}: Props): JSX.Element {
  const containerRef = React.useRef(null);
  const inputRef = React.useRef(null);

  const onValidate = () => {
    if (!confirmValue) {
      onConfirmed();
      onClose();
    } else {
      const input = inputRef.current;
      if (confirmValue === input.value) {
        onConfirmed();
        onClose();
      } else {
        const container = containerRef.current;
        input.value = '';
        container.classList.remove('animate__animated', 'animate__bounceIn');
        container.offsetWidth; // some magic: https://css-tricks.com/restart-css-animation/
        container.classList.add('animate__animated', 'animate__bounceIn');
      }
    }
  };

  const handleKeyDown = (e): void => {
    if (e.keyCode === keyCodeConstants.KEY_RETURN) {
      onValidate();
    }
    e.stopPropagation();
  };

  const renderButtons = (): JSX.Element => {
    const defaultButtons = [
      {
        label: LANG.ok,
        className: 'btn-default primary',
        onClick: () => {
          onValidate();
        },
      },
      {
        label: LANG.cancel,
        className: 'btn-default',
        onClick: () => {
          onCancel();
          onClose();
        },
      },
    ];
    return <ButtonGroup className="btn-right" buttons={defaultButtons} />;
  };

  return (
    <Modal>
      <div className={classNames('confirm-prompt-dialog-container', 'animate__animated', 'animate__bounceIn')} ref={containerRef}>
        <div className="caption">{caption}</div>
        <pre className="message">{message}</pre>
        <input
          autoFocus
          ref={inputRef}
          className="text-input"
          type="text"
          onKeyDown={handleKeyDown}
          placeholder={confirmValue}
        />
        <div className="footer">
          {renderButtons()}
        </div>
      </div>
    </Modal>
  );
}

export default ConfirmPrompt;
