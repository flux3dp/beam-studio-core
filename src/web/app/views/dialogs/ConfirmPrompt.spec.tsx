/* eslint-disable import/first */
import * as React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

jest.mock('helpers/i18n', () => ({
  lang: {
    alert: {
      ok: 'OK',
      cancel: 'Cancel',
    },
  },
}));

import ConfirmPrompt from './ConfirmPrompt';

describe('should render correctly', () => {
  test('no confirmValue', () => {
    const onConfirmed = jest.fn();
    const onClose = jest.fn();
    const onCancel = jest.fn();
    const wrapper = mount(<ConfirmPrompt
      caption="POWER TOO HIGH"
      message={'Using lower laser power (under 70%) will extends laser tube\'s lifetime.\nEnter "NOTED" to proceed.'}
      confirmValue=""
      onConfirmed={onConfirmed}
      onClose={onClose}
      onCancel={onCancel}
    />);
    expect(toJson(wrapper)).toMatchSnapshot();

    wrapper.find('button.btn-default').at(0).simulate('click');
    expect(onConfirmed).toHaveBeenCalledTimes(1);
    expect(onClose).toHaveBeenCalledTimes(1);

    jest.resetAllMocks();
    wrapper.find('button.btn-default').at(1).simulate('click');
    expect(onCancel).toHaveBeenCalledTimes(1);
    expect(onClose).toHaveBeenCalledTimes(1);

    jest.resetAllMocks();
    const stopPropagation = jest.fn();
    wrapper.find('input').simulate('keydown', {
      keyCode: 13,
      stopPropagation,
    });
    expect(onConfirmed).toHaveBeenCalledTimes(1);
    expect(stopPropagation).toHaveBeenCalledTimes(1);

    jest.resetAllMocks();
    wrapper.find('input').simulate('keydown', {
      keyCode: 14,
      stopPropagation,
    });
    expect(onConfirmed).not.toHaveBeenCalled();
    expect(onClose).not.toHaveBeenCalled();
    expect(stopPropagation).toHaveBeenCalledTimes(1);
  });

  test('has confirmValue', () => {
    const onConfirmed = jest.fn();
    const onClose = jest.fn();
    const onCancel = jest.fn();
    const wrapper = mount(<ConfirmPrompt
      caption="POWER TOO HIGH"
      message={'Using lower laser power (under 70%) will extends laser tube\'s lifetime.\nEnter "NOTED" to proceed.'}
      confirmValue="NOTED"
      onConfirmed={onConfirmed}
      onClose={onClose}
      onCancel={onCancel}
    />);
    expect(toJson(wrapper)).toMatchSnapshot();

    wrapper.find('input').instance().value = 'NOTED';
    wrapper.find('button.btn-default').at(0).simulate('click');
    expect(onConfirmed).toHaveBeenCalledTimes(1);
    expect(onClose).toHaveBeenCalledTimes(1);

    jest.resetAllMocks();
    wrapper.find('input').instance().value = 'Hello';
    wrapper.find('button.btn-default').at(0).simulate('click');
    expect(wrapper.find('input').instance().value).toBe('');
    expect(onConfirmed).not.toHaveBeenCalled();
    expect(onClose).not.toHaveBeenCalled();
  });
});
