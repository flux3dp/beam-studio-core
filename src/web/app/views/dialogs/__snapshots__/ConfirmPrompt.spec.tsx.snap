// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly has confirmValue 1`] = `
<ConfirmPrompt
  caption="POWER TOO HIGH"
  confirmValue="NOTED"
  message="Using lower laser power (under 70%) will extends laser tube's lifetime.
Enter \\"NOTED\\" to proceed."
  onCancel={[MockFunction]}
  onClose={[MockFunction]}
  onConfirmed={[MockFunction]}
>
  <View>
    <div
      className="modal-window"
    >
      <div
        className="modal-background"
        onClick={[Function]}
      />
      <div
        className="modal-body"
      >
        <div
          className="confirm-prompt-dialog-container animate__animated animate__bounceIn"
        >
          <div
            className="caption"
          >
            POWER TOO HIGH
          </div>
          <pre
            className="message"
          >
            Using lower laser power (under 70%) will extends laser tube's lifetime.
Enter "NOTED" to proceed.
          </pre>
          <input
            autoFocus={true}
            className="text-input"
            id="confirm-prompt-input"
            onKeyDown={[Function]}
            placeholder="NOTED"
            type="text"
          />
          <div
            className="footer"
          >
            <ButtonGroup
              buttons={
                Array [
                  Object {
                    "className": "btn-default primary",
                    "label": "OK",
                    "onClick": [Function],
                  },
                  Object {
                    "className": "btn-default",
                    "label": "Cancel",
                    "onClick": [Function],
                  },
                ]
              }
              className="btn-right"
            >
              <div
                className="button-group btn-right"
              >
                <button
                  className="btn btn-default primary"
                  dangerouslySetInnerHTML={
                    Object {
                      "__html": "OK",
                    }
                  }
                  data-test-key="ok"
                  key="0"
                  onClick={[Function]}
                  type="button"
                />
                <button
                  className="btn btn-default"
                  dangerouslySetInnerHTML={
                    Object {
                      "__html": "Cancel",
                    }
                  }
                  data-test-key="cancel"
                  key="1"
                  onClick={[Function]}
                  type="button"
                />
              </div>
            </ButtonGroup>
          </div>
        </div>
      </div>
    </div>
  </View>
</ConfirmPrompt>
`;

exports[`should render correctly no confirmValue 1`] = `
<ConfirmPrompt
  caption="POWER TOO HIGH"
  confirmValue=""
  message="Using lower laser power (under 70%) will extends laser tube's lifetime.
Enter \\"NOTED\\" to proceed."
  onCancel={[MockFunction]}
  onClose={[MockFunction]}
  onConfirmed={[MockFunction]}
>
  <View>
    <div
      className="modal-window"
    >
      <div
        className="modal-background"
        onClick={[Function]}
      />
      <div
        className="modal-body"
      >
        <div
          className="confirm-prompt-dialog-container animate__animated animate__bounceIn"
        >
          <div
            className="caption"
          >
            POWER TOO HIGH
          </div>
          <pre
            className="message"
          >
            Using lower laser power (under 70%) will extends laser tube's lifetime.
Enter "NOTED" to proceed.
          </pre>
          <input
            autoFocus={true}
            className="text-input"
            id="confirm-prompt-input"
            onKeyDown={[Function]}
            placeholder=""
            type="text"
          />
          <div
            className="footer"
          >
            <ButtonGroup
              buttons={
                Array [
                  Object {
                    "className": "btn-default primary",
                    "label": "OK",
                    "onClick": [Function],
                  },
                  Object {
                    "className": "btn-default",
                    "label": "Cancel",
                    "onClick": [Function],
                  },
                ]
              }
              className="btn-right"
            >
              <div
                className="button-group btn-right"
              >
                <button
                  className="btn btn-default primary"
                  dangerouslySetInnerHTML={
                    Object {
                      "__html": "OK",
                    }
                  }
                  data-test-key="ok"
                  key="0"
                  onClick={[Function]}
                  type="button"
                />
                <button
                  className="btn btn-default"
                  dangerouslySetInnerHTML={
                    Object {
                      "__html": "Cancel",
                    }
                  }
                  data-test-key="cancel"
                  key="1"
                  onClick={[Function]}
                  type="button"
                />
              </div>
            </ButtonGroup>
          </div>
        </div>
      </div>
    </div>
  </View>
</ConfirmPrompt>
`;
